File: tests/process_data_test.cpp
#include "jetbus/process_data.hpp"

#include <catch2/catch_approx.hpp>
#include <catch2/catch_test_macros.hpp>

#include <unordered_map>
#include <string>

TEST_CASE("ProcessData decodes status bits and weight values", "[process_data]") {
    jetbus::ProcessData process_data;
    std::unordered_map<std::string, std::string> cache;

    cache[jetbus::commands::cia461_decimals().path] = "1";
    cache[jetbus::commands::cia461_unit().path] = std::to_string(0x00020000);
    cache[jetbus::commands::cia461_net_value().path] = "123";
    cache[jetbus::commands::cia461_gross_value().path] = "456";
    cache[jetbus::commands::cia461_tare_value().path] = "30";
    cache[jetbus::commands::imd_application_mode().path] = "0";

    int status_value = 0;
    status_value |= (1 << 6);  // manual tare
    status_value |= (1 << 11); // center of zero
    cache[jetbus::commands::cia461_weight_status().path] = std::to_string(status_value);

    process_data.update(cache);

    REQUIRE(process_data.decimals() == 1);
    REQUIRE(process_data.unit() == "kg");
    REQUIRE(process_data.weight().net == Catch::Approx(12.3));
    REQUIRE(process_data.weight().gross == Catch::Approx(45.6));
    REQUIRE(process_data.weight().tare == Catch::Approx(3.0));
    REQUIRE(process_data.tare_mode() == dse::TareMode::Tare);
    REQUIRE(process_data.weight_stable());
    REQUIRE(process_data.center_of_zero());
    REQUIRE_FALSE(process_data.zero_required());
    REQUIRE(process_data.legal_for_trade());
    REQUIRE_FALSE(process_data.underload());
    REQUIRE_FALSE(process_data.overload());
    REQUIRE_FALSE(process_data.higher_safe_load_limit());
    REQUIRE_FALSE(process_data.general_scale_error());
    REQUIRE_FALSE(process_data.scale_alarm());
}

File: tests/processData.test.ts
import { describe, expect, it } from "vitest";
import { ProcessData } from "../src/jetbus/processData";
import { commands } from "../src/jetbus/commands";
import { TareMode } from "../src/dse/types";

describe("ProcessData", () => {
  it("interprets cached values correctly", () => {
    const cache = new Map<string, string>();

    // Status base value
    let weightStatus = 0;
    weightStatus |= 1 << 1; // scale alarm
    weightStatus |= 2 << 2; // limit status -> overload
    weightStatus |= 1 << 6; // manual tare present
    weightStatus |= 0 << 7; // weight type manual
    weightStatus |= 1 << 10; // zero required
    weightStatus |= 1 << 11; // center of zero
    weightStatus |= 0 << 12; // inside zero false

    cache.set(commands.cia461WeightStatus().path, weightStatus.toString());
    cache.set(commands.cia461Decimals().path, "2");
    cache.set(commands.cia461Unit().path, (0x00020000).toString(10)); // kg
    cache.set(commands.cia461NetValue().path, "12345");
    cache.set(commands.cia461GrossValue().path, "15000");
    cache.set(commands.cia461TareValue().path, "2655");
    cache.set(commands.imdApplicationMode().path, "0");

    const processData = new ProcessData();
    processData.update(cache);

    expect(processData.weight().net).toBeCloseTo(123.45, 6);
    expect(processData.weight().gross).toBeCloseTo(150.0, 6);
    expect(processData.weight().tare).toBeCloseTo(26.55, 6);
    expect(processData.printableWeight().net).toBe("123.45");
    expect(processData.unit()).toBe("kg");
    expect(processData.decimals()).toBe(2);
    expect(processData.tareMode()).toBe(TareMode.Tare);
    expect(processData.weightStable()).toBe(true);
    expect(processData.zeroRequired()).toBe(true);
    expect(processData.centerOfZero()).toBe(true);
    expect(processData.insideZero()).toBe(false);
    expect(processData.legalForTrade()).toBe(true);
    expect(processData.overload()).toBe(true);
    expect(processData.underload()).toBe(false);
    expect(processData.generalScaleError()).toBe(false);
    expect(processData.scaleAlarm()).toBe(true);
  });
});
